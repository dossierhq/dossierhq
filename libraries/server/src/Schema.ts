import type {
  AdminSchemaSpecification,
  AdminSchemaSpecificationUpdate,
  ErrorType,
  PromiseResult,
  SchemaSpecificationUpdatePayload,
} from '@jonasb/datadata-core';
import { AdminSchema, ok } from '@jonasb/datadata-core';
import type { DatabaseAdapter, TransactionContext } from '@jonasb/datadata-database-adapter';

export async function getSchemaSpecification(
  databaseAdapter: DatabaseAdapter,
  context: TransactionContext
): PromiseResult<AdminSchemaSpecification, ErrorType.Generic> {
  const { logger } = context;
  logger.info('Loading schema');
  const result = await databaseAdapter.schemaGetSpecification(context);
  if (result.isError()) {
    return result;
  }

  const specification = result.value;
  if (!specification) {
    logger.info('No schema set, defaulting to empty');
    return ok({ entityTypes: [], valueTypes: [] });
  }
  logger.info(
    'Loaded schema with %d entity types and  %d value types',
    specification.entityTypes.length,
    specification.valueTypes.length
  );
  return ok(specification);
}

export async function updateSchemaSpecification(
  databaseAdapter: DatabaseAdapter,
  context: TransactionContext,
  schemaSpec: AdminSchemaSpecificationUpdate
): PromiseResult<SchemaSpecificationUpdatePayload, ErrorType.BadRequest | ErrorType.Generic> {
  return await context.withTransaction(async (context) => {
    const previousSpecificationResult = await getSchemaSpecification(databaseAdapter, context);
    if (previousSpecificationResult.isError()) {
      return previousSpecificationResult;
    }
    const schema = new AdminSchema(previousSpecificationResult.value);
    const mergeResult = schema.mergeWith(schemaSpec);
    if (mergeResult.isError()) {
      return mergeResult;
    }
    const schemaSpecification = mergeResult.value;
    // TODO return with 'none' if same as previous schema
    const updateResult = await databaseAdapter.schemaUpdateSpecification(
      context,
      schemaSpecification
    );
    if (updateResult.isError()) {
      return updateResult;
    }
    return ok({ effect: 'updated', schemaSpecification });
  });
}
