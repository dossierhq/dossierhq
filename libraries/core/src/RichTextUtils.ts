import { RichTextNodeType } from './Schema.js';
import type {
  HeadingTagType,
  SerializedHeadingNode,
  SerializedParagraphNode,
  SerializedTextNode,
  TextFormatType,
} from './third-party/Lexical.js';
import { TEXT_TYPE_TO_FORMAT } from './third-party/Lexical.js';
import type {
  EntityReference,
  RichText,
  RichTextEntityLinkNode,
  RichTextEntityNode,
  RichTextNode,
  RichTextValueItemNode,
  ValueItem,
} from './Types.js';

export function createRichTextRootNode(children: RichTextNode[]): RichText {
  return {
    root: {
      direction: 'ltr',
      format: '',
      indent: 0,
      type: 'root',
      version: 1,
      children,
    },
  };
}

export function createRichTextHeadingNode(
  tag: HeadingTagType,
  children: RichTextNode[]
): SerializedHeadingNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    tag,
    type: 'heading',
    version: 1,
    children,
  };
}

export function createRichTextParagraphNode(children: RichTextNode[]): SerializedParagraphNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'paragraph',
    version: 1,
    children,
  };
}

export function createRichTextTextNode(
  text: string,
  options?: { format?: TextFormatType[] }
): SerializedTextNode {
  let formatValue = 0;
  if (options?.format) {
    for (const formatItem of options.format) {
      const formatFlag = TEXT_TYPE_TO_FORMAT[formatItem];
      if (formatFlag !== undefined) {
        formatValue |= formatFlag;
      }
    }
  }
  return {
    detail: 0,
    format: formatValue,
    mode: 'normal',
    style: '',
    text,
    type: 'text',
    version: 1,
  };
}

export function createRichTextEntityNode(reference: EntityReference): RichTextEntityNode {
  return { type: RichTextNodeType.entity, reference, version: 1 };
}

export function createRichTextEntityLinkNode(
  reference: EntityReference,
  children: RichTextNode[]
): RichTextEntityLinkNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: RichTextNodeType.entityLink,
    reference,
    version: 1,
    children,
  };
}

export function createRichTextValueItemNode<T extends ValueItem<string, object>>(
  data: T
): RichTextValueItemNode {
  return { type: RichTextNodeType.valueItem, data, version: 1 };
}
