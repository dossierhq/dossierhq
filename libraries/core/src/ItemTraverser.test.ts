import { describe, expect, test } from 'vitest';
import type { AdminItemTraverseNode } from './ItemTraverser.js';
import {
  AdminItemTraverseNodeType,
  traverseAdminEntity,
  traverseAdminValueItem,
} from './ItemTraverser.js';
import { visitorPathToString } from './ItemUtils.js';
import { createRichTextRootNode, createRichTextValueItemNode } from './RichTextUtils.js';
import { AdminSchema, FieldType } from './Schema.js';

const schema = new AdminSchema({
  entityTypes: [
    {
      name: 'Foo',
      adminOnly: false,
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'stringList', type: FieldType.String, list: true },
        { name: 'twoStrings', type: FieldType.ValueType, valueTypes: ['TwoStrings'] },
        { name: 'richText', type: FieldType.RichText },
      ],
    },
  ],
  valueTypes: [
    {
      name: 'TwoStrings',
      adminOnly: false,
      fields: [
        { name: 'string1', type: FieldType.String },
        { name: 'string2', type: FieldType.String },
      ],
    },
  ],
});

function collectTraverseNodes(generator: Generator<AdminItemTraverseNode>) {
  const result = [];
  for (const node of generator) {
    const path = visitorPathToString(node.path);
    switch (node.type) {
      case AdminItemTraverseNodeType.error:
        result.push({ type: node.type, path, message: node.message });
        break;
      case AdminItemTraverseNodeType.field:
        result.push({ type: node.type, path, value: node.value });
        break;
      case AdminItemTraverseNodeType.valueItem:
        result.push({ type: node.type, path, valueItem: node.valueItem });
        break;
      default: {
        result.push(node);
      }
    }
  }
  return result;
}

describe('traverseAdminEntity', () => {
  test('Empty Foo entity', () => {
    const nodes = collectTraverseNodes(
      traverseAdminEntity(schema, ['entity'], { info: { type: 'Foo' }, fields: {} })
    );
    expect(nodes).toMatchInlineSnapshot(`
      [
        {
          "path": "entity.fields.string",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "string",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "string",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.stringList",
          "type": "field",
          "value": undefined,
        },
        {
          "path": "entity.fields.twoStrings",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "twoStrings",
            "type": "ValueType",
            "valueTypes": [
              "TwoStrings",
            ],
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.richText",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "path": [
            "entity",
            "fields",
            "richText",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
      ]
    `);
  });

  test('Foo with two strings in list', () => {
    const nodes = collectTraverseNodes(
      traverseAdminEntity(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: { stringList: ['string1', 'string2'] },
      })
    );
    expect(nodes).toMatchInlineSnapshot(`
      [
        {
          "path": "entity.fields.string",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "string",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "string",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.stringList",
          "type": "field",
          "value": [
            "string1",
            "string2",
          ],
        },
        {
          "fieldSpec": {
            "list": true,
            "name": "stringList",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "stringList",
            0,
          ],
          "type": "fieldItem",
          "value": "string1",
        },
        {
          "fieldSpec": {
            "list": true,
            "name": "stringList",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "stringList",
            1,
          ],
          "type": "fieldItem",
          "value": "string2",
        },
        {
          "path": "entity.fields.twoStrings",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "twoStrings",
            "type": "ValueType",
            "valueTypes": [
              "TwoStrings",
            ],
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.richText",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "path": [
            "entity",
            "fields",
            "richText",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
      ]
    `);
  });

  test('Foo entity with TwoStrings value item', () => {
    const nodes = collectTraverseNodes(
      traverseAdminEntity(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: {
          string: 'string1',
          stringList: ['string2.1', 'string2.2'],
          twoStrings: { type: 'TwoStrings', string1: 'two-1', string2: 'two-2' },
        },
      })
    );
    expect(nodes).toMatchInlineSnapshot(`
      [
        {
          "path": "entity.fields.string",
          "type": "field",
          "value": "string1",
        },
        {
          "fieldSpec": {
            "name": "string",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "string",
          ],
          "type": "fieldItem",
          "value": "string1",
        },
        {
          "path": "entity.fields.stringList",
          "type": "field",
          "value": [
            "string2.1",
            "string2.2",
          ],
        },
        {
          "fieldSpec": {
            "list": true,
            "name": "stringList",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "stringList",
            0,
          ],
          "type": "fieldItem",
          "value": "string2.1",
        },
        {
          "fieldSpec": {
            "list": true,
            "name": "stringList",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "stringList",
            1,
          ],
          "type": "fieldItem",
          "value": "string2.2",
        },
        {
          "path": "entity.fields.twoStrings",
          "type": "field",
          "value": {
            "string1": "two-1",
            "string2": "two-2",
            "type": "TwoStrings",
          },
        },
        {
          "fieldSpec": {
            "name": "twoStrings",
            "type": "ValueType",
            "valueTypes": [
              "TwoStrings",
            ],
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
          ],
          "type": "fieldItem",
          "value": {
            "string1": "two-1",
            "string2": "two-2",
            "type": "TwoStrings",
          },
        },
        {
          "path": "entity.fields.twoStrings",
          "type": "valueItem",
          "valueItem": {
            "string1": "two-1",
            "string2": "two-2",
            "type": "TwoStrings",
          },
        },
        {
          "path": "entity.fields.twoStrings.string1",
          "type": "field",
          "value": "two-1",
        },
        {
          "fieldSpec": {
            "name": "string1",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
            "string1",
          ],
          "type": "fieldItem",
          "value": "two-1",
        },
        {
          "path": "entity.fields.twoStrings.string2",
          "type": "field",
          "value": "two-2",
        },
        {
          "fieldSpec": {
            "name": "string2",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
            "string2",
          ],
          "type": "fieldItem",
          "value": "two-2",
        },
        {
          "path": "entity.fields.richText",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "path": [
            "entity",
            "fields",
            "richText",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
      ]
    `);
  });

  test('Foo entity with rich text with TwoStrings value item', () => {
    const nodes = collectTraverseNodes(
      traverseAdminEntity(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: {
          richText: createRichTextRootNode([
            createRichTextValueItemNode({ type: 'TwoStrings', string1: 'two-1', string2: 'two-2' }),
          ]),
        },
      })
    );
    expect(nodes).toMatchInlineSnapshot(`
      [
        {
          "path": "entity.fields.string",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "string",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "string",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.stringList",
          "type": "field",
          "value": undefined,
        },
        {
          "path": "entity.fields.twoStrings",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "twoStrings",
            "type": "ValueType",
            "valueTypes": [
              "TwoStrings",
            ],
          },
          "path": [
            "entity",
            "fields",
            "twoStrings",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "entity.fields.richText",
          "type": "field",
          "value": {
            "root": {
              "children": [
                {
                  "data": {
                    "string1": "two-1",
                    "string2": "two-2",
                    "type": "TwoStrings",
                  },
                  "type": "valueItem",
                  "version": 1,
                },
              ],
              "direction": "ltr",
              "format": "",
              "indent": 0,
              "type": "root",
              "version": 1,
            },
          },
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "path": [
            "entity",
            "fields",
            "richText",
          ],
          "type": "fieldItem",
          "value": {
            "root": {
              "children": [
                {
                  "data": {
                    "string1": "two-1",
                    "string2": "two-2",
                    "type": "TwoStrings",
                  },
                  "type": "valueItem",
                  "version": 1,
                },
              ],
              "direction": "ltr",
              "format": "",
              "indent": 0,
              "type": "root",
              "version": 1,
            },
          },
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "node": {
            "children": [
              {
                "data": {
                  "string1": "two-1",
                  "string2": "two-2",
                  "type": "TwoStrings",
                },
                "type": "valueItem",
                "version": 1,
              },
            ],
            "direction": "ltr",
            "format": "",
            "indent": 0,
            "type": "root",
            "version": 1,
          },
          "path": [
            "entity",
            "fields",
            "richText",
          ],
          "type": "richTextNode",
        },
        {
          "fieldSpec": {
            "name": "richText",
            "type": "RichText",
          },
          "node": {
            "data": {
              "string1": "two-1",
              "string2": "two-2",
              "type": "TwoStrings",
            },
            "type": "valueItem",
            "version": 1,
          },
          "path": [
            "entity",
            "fields",
            "richText",
            0,
          ],
          "type": "richTextNode",
        },
        {
          "path": "entity.fields.richText[0].data",
          "type": "valueItem",
          "valueItem": {
            "string1": "two-1",
            "string2": "two-2",
            "type": "TwoStrings",
          },
        },
        {
          "path": "entity.fields.richText[0].data.string1",
          "type": "field",
          "value": "two-1",
        },
        {
          "fieldSpec": {
            "name": "string1",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "richText",
            0,
            "data",
            "string1",
          ],
          "type": "fieldItem",
          "value": "two-1",
        },
        {
          "path": "entity.fields.richText[0].data.string2",
          "type": "field",
          "value": "two-2",
        },
        {
          "fieldSpec": {
            "name": "string2",
            "type": "String",
          },
          "path": [
            "entity",
            "fields",
            "richText",
            0,
            "data",
            "string2",
          ],
          "type": "fieldItem",
          "value": "two-2",
        },
      ]
    `);
  });
});

describe('traverseAdminValueItem', () => {
  test('Empty TwoStrings value item', () => {
    const nodes = collectTraverseNodes(
      traverseAdminValueItem(schema, ['valueItem'], { type: 'TwoStrings' })
    );
    expect(nodes).toMatchInlineSnapshot(`
      [
        {
          "path": "valueItem",
          "type": "valueItem",
          "valueItem": {
            "type": "TwoStrings",
          },
        },
        {
          "path": "valueItem.string1",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "string1",
            "type": "String",
          },
          "path": [
            "valueItem",
            "string1",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
        {
          "path": "valueItem.string2",
          "type": "field",
          "value": undefined,
        },
        {
          "fieldSpec": {
            "name": "string2",
            "type": "String",
          },
          "path": [
            "valueItem",
            "string2",
          ],
          "type": "fieldItem",
          "value": undefined,
        },
      ]
    `);
  });
});
