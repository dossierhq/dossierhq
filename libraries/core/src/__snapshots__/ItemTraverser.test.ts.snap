// Vitest Snapshot v1

exports[`traverseEntity > Empty Foo entity 1`] = `
[
  {
    "path": "entity.fields.string",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "string",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.stringList",
    "type": "field",
    "value": undefined,
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "list": false,
      "name": "twoStrings",
      "required": false,
      "type": "ValueItem",
      "valueTypes": [
        "TwoStrings",
      ],
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.richText",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.adminOnlyString",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": true,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "adminOnlyString",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "adminOnlyString",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;

exports[`traverseEntity > Foo entity with TwoStrings value item 1`] = `
[
  {
    "path": "entity.fields.string",
    "type": "field",
    "value": "string1",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "string",
    ],
    "type": "fieldItem",
    "value": "string1",
  },
  {
    "path": "entity.fields.stringList",
    "type": "field",
    "value": [
      "string2.1",
      "string2.2",
    ],
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": true,
      "matchPattern": null,
      "multiline": false,
      "name": "stringList",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "stringList",
      0,
    ],
    "type": "fieldItem",
    "value": "string2.1",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": true,
      "matchPattern": null,
      "multiline": false,
      "name": "stringList",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "stringList",
      1,
    ],
    "type": "fieldItem",
    "value": "string2.2",
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "field",
    "value": {
      "string1": "two-1",
      "string2": "two-2",
      "type": "TwoStrings",
    },
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "list": false,
      "name": "twoStrings",
      "required": false,
      "type": "ValueItem",
      "valueTypes": [
        "TwoStrings",
      ],
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
    ],
    "type": "fieldItem",
    "value": {
      "string1": "two-1",
      "string2": "two-2",
      "type": "TwoStrings",
    },
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "valueItem",
    "valueItem": {
      "string1": "two-1",
      "string2": "two-2",
      "type": "TwoStrings",
    },
  },
  {
    "path": "entity.fields.twoStrings.string1",
    "type": "field",
    "value": "two-1",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string1",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
      "string1",
    ],
    "type": "fieldItem",
    "value": "two-1",
  },
  {
    "path": "entity.fields.twoStrings.string2",
    "type": "field",
    "value": "two-2",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string2",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
      "string2",
    ],
    "type": "fieldItem",
    "value": "two-2",
  },
  {
    "path": "entity.fields.richText",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.adminOnlyString",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": true,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "adminOnlyString",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "adminOnlyString",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;

exports[`traverseEntity > Foo entity with rich text with TwoStrings value item 1`] = `
[
  {
    "path": "entity.fields.string",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "string",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.stringList",
    "type": "field",
    "value": undefined,
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "list": false,
      "name": "twoStrings",
      "required": false,
      "type": "ValueItem",
      "valueTypes": [
        "TwoStrings",
      ],
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.richText",
    "type": "field",
    "value": {
      "root": {
        "children": [
          {
            "data": {
              "string1": "two-1",
              "string2": "two-2",
              "type": "TwoStrings",
            },
            "format": "",
            "type": "valueItem",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "type": "root",
        "version": 1,
      },
    },
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "fieldItem",
    "value": {
      "root": {
        "children": [
          {
            "data": {
              "string1": "two-1",
              "string2": "two-2",
              "type": "TwoStrings",
            },
            "format": "",
            "type": "valueItem",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "type": "root",
        "version": 1,
      },
    },
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "node": {
      "children": [
        {
          "data": {
            "string1": "two-1",
            "string2": "two-2",
            "type": "TwoStrings",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "richTextNode",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "node": {
      "data": {
        "string1": "two-1",
        "string2": "two-2",
        "type": "TwoStrings",
      },
      "format": "",
      "type": "valueItem",
      "version": 1,
    },
    "path": [
      "entity",
      "fields",
      "richText",
      0,
    ],
    "type": "richTextNode",
  },
  {
    "path": "entity.fields.richText[0].data",
    "type": "valueItem",
    "valueItem": {
      "string1": "two-1",
      "string2": "two-2",
      "type": "TwoStrings",
    },
  },
  {
    "path": "entity.fields.richText[0].data.string1",
    "type": "field",
    "value": "two-1",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string1",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "richText",
      0,
      "data",
      "string1",
    ],
    "type": "fieldItem",
    "value": "two-1",
  },
  {
    "path": "entity.fields.richText[0].data.string2",
    "type": "field",
    "value": "two-2",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string2",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "richText",
      0,
      "data",
      "string2",
    ],
    "type": "fieldItem",
    "value": "two-2",
  },
  {
    "path": "entity.fields.adminOnlyString",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": true,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "adminOnlyString",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "adminOnlyString",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;

exports[`traverseEntity > Foo with adminOnly field (published) 1`] = `
[
  {
    "path": "entity.fields.string",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "string",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.stringList",
    "type": "field",
    "value": undefined,
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "list": false,
      "name": "twoStrings",
      "required": false,
      "type": "ValueItem",
      "valueTypes": [
        "TwoStrings",
      ],
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.richText",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;

exports[`traverseEntity > Foo with two strings in list 1`] = `
[
  {
    "path": "entity.fields.string",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "string",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.stringList",
    "type": "field",
    "value": [
      "string1",
      "string2",
    ],
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": true,
      "matchPattern": null,
      "multiline": false,
      "name": "stringList",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "stringList",
      0,
    ],
    "type": "fieldItem",
    "value": "string1",
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": true,
      "matchPattern": null,
      "multiline": false,
      "name": "stringList",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "stringList",
      1,
    ],
    "type": "fieldItem",
    "value": "string2",
  },
  {
    "path": "entity.fields.twoStrings",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "list": false,
      "name": "twoStrings",
      "required": false,
      "type": "ValueItem",
      "valueTypes": [
        "TwoStrings",
      ],
    },
    "path": [
      "entity",
      "fields",
      "twoStrings",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.richText",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "entityTypes": [],
      "linkEntityTypes": [],
      "list": false,
      "name": "richText",
      "required": false,
      "richTextNodes": [],
      "type": "RichText",
      "valueTypes": [],
    },
    "path": [
      "entity",
      "fields",
      "richText",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "entity.fields.adminOnlyString",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": true,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "adminOnlyString",
      "required": false,
      "type": "String",
    },
    "path": [
      "entity",
      "fields",
      "adminOnlyString",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;

exports[`traverseValueItem > Empty TwoStrings value item 1`] = `
[
  {
    "path": "valueItem",
    "type": "valueItem",
    "valueItem": {
      "type": "TwoStrings",
    },
  },
  {
    "path": "valueItem.string1",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string1",
      "required": false,
      "type": "String",
    },
    "path": [
      "valueItem",
      "string1",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
  {
    "path": "valueItem.string2",
    "type": "field",
    "value": undefined,
  },
  {
    "fieldSpec": {
      "adminOnly": false,
      "index": null,
      "list": false,
      "matchPattern": null,
      "multiline": false,
      "name": "string2",
      "required": false,
      "type": "String",
    },
    "path": [
      "valueItem",
      "string2",
    ],
    "type": "fieldItem",
    "value": undefined,
  },
]
`;
